vmm/src/serial_buffer.rs

```diff
@@ -3,173 +3,95 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-use crate::serial_manager::EpollDispatch;
-
-use std::io::Write;
-use std::os::unix::io::RawFd;
+use std::{
+    collections::VecDeque,
+    io::Write,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
 
 // Circular buffer implementation for serial output.
 // Read from head; push to tail
 pub(crate) struct SerialBuffer {
-    buffer: Vec<u8>,
-    head: usize,
-    tail: usize,
+    buffer: VecDeque<u8>,
     out: Box<dyn Write + Send>,
-    buffering: bool,
-    out_fd: Option<RawFd>,
-    epoll_fd: Option<RawFd>,
+    write_out: Arc<AtomicBool>,
 }
 
-const MAX_BUFFER_SIZE: usize = 16 << 10;
-
 impl SerialBuffer {
-    pub(crate) fn new(out: Box<dyn Write + Send>) -> Self {
+    pub(crate) fn new(out: Box<dyn Write + Send>, write_out: Arc<AtomicBool>) -> Self {
         Self {
-            buffer: vec![],
-            head: 0,
-            tail: 0,
+            buffer: VecDeque::new(),
             out,
-            buffering: false,
-            out_fd: None,
-            epoll_fd: None,
+            write_out,
         }
     }
+}
 
-    pub(crate) fn add_out_fd(&mut self, out_fd: RawFd) {
-        self.out_fd = Some(out_fd);
-    }
+impl Write for SerialBuffer {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+        // Simply fill the buffer if we're not allowed to write to the out
+        // device.
+        if !self.write_out.load(Ordering::Acquire) {
+            self.buffer.extend(buf);
+            return Ok(buf.len());
+        }
 
-    pub(crate) fn add_epoll_fd(&mut self, epoll_fd: RawFd) {
-        self.epoll_fd = Some(epoll_fd);
-    }
+        // In case we're allowed to write to the out device, we flush the
+        // content of the buffer.
+        self.flush()?;
+
+        // If after flushing the buffer, it's still not empty, that means
+        // something wrong happened during the write, and we should fill
+        // the buffer with what's coming from the serial.
```

Comment:
Rather than something went wrong, surely it was more that only a subset of the bytes was written?